authz/grpc_audit_end2end_test.go

```diff
@@ -0,0 +1,332 @@
+/*
+ *
+ * Copyright 2023 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package authz_test
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"io"
+	"net"
+	"os"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/authz"
+	"google.golang.org/grpc/authz/audit"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/internal/grpctest"
+	testgrpc "google.golang.org/grpc/interop/grpc_testing"
+	testpb "google.golang.org/grpc/interop/grpc_testing"
+	"google.golang.org/grpc/testdata"
+
+	_ "google.golang.org/grpc/authz/audit/stdout"
+)
+
+func TestAudit(t *testing.T) {
+	grpctest.RunSubTests(t, s{})
+}
+
+type statAuditLogger struct {
+	AuthzDescisionStat map[bool]int      // Map to hold the counts of authorization decisions
+	EventContent       map[string]string // Map to hold event fields in key:value fashion
+	SpiffeIds          []string          // Slice to hold collected SPIFFE IDs
+}
+
+func (s *statAuditLogger) Log(event *audit.Event) {
+	s.AuthzDescisionStat[event.Authorized]++
+	s.EventContent["rpc_method"] = event.FullMethodName
+	s.EventContent["principal"] = event.Principal
+	s.EventContent["policy_name"] = event.PolicyName
+	s.EventContent["matched_rule"] = event.MatchedRule
+	s.EventContent["authorized"] = strconv.FormatBool(event.Authorized)
```

Comment:
Why not `s.EventContent = event` and make this field type `*audit.Event`?  (Also consider a rename to `lastEvent`?)

Reply:
Here EventContent is a map created at `loggerBuilder` but manipulated by `statAuditLogger`. At the end I'm using `loggerBuilder` to access it's content. I'm not sure how to achieve the same using `*audit.Event` - maybe only if I use a slice or map wrapping it?

Reply:
I DON"T UNDERSTAND THIS>  Why can't you just change the type of this field (now `lastEventContent`) to `*audit.Event` and rename the field to `lastEvent`?  Are you saying the tricky bit is communicating between the test and the constructed logger?  If so, 2 options:

1. Use `*audit.Event` and `*s.lastEvent = *event`
2. Use `**audit.Event` and `*s.lastEvent = event`

Reply:
Wow what happened with my last comment here??  Sometimes the shift key sticks on my keyboard, but I can't believe I didn't see it.  Sorry if that came off as rude, I didn't mean to type in all-caps.

Reply:
No worries!